NMS-14281: Add instructions for managing configuration files with git (#4962)
OpenNMS/opennms: docs/modules/deployment/pages/upgrade/git-diff.adoc

[[git-diff]]
= Manage Configuration Changes with Git

This tutorial was originally written on the OpenNMS Discourse forum (see https://opennms.discourse.group/t/managing-configuration-changes-with-git/2327[Managing Configuration Changes with Git]).

> If you want to manage your configuration files with git but haven't used it for version control before, GitHub's https://docs.github.com/en/get-started/using-git[Using Git] tutorial covers the basics.

Reviewer: Capitalize when referring to the software. Lowercase when showing a command.

You can also read https://opennms.discourse.group/t/developing-with-git/2228[Developing with Git] on the OpenNMS Discourse forum.

If you have an existing OpenNMS {page-component-title} installation, check for the `etc-pristine` directory.
If your {page-component-title} environment does not have this directory, you will need to upgrade to a version that does.

IMPORTANT: Remember to stop {page-component-title} before changing files in the `etc-pristine` directory.

== Start with pristine `etc` directory

[{tabs}]
====
Debian OSes::
+
--
IMPORTANT: `$OPENNMS_HOME/etc` is linked to `/etc/opennms`.
Do not move that folder.
Moving the folder will break the link and will cause future upgrades to fail.

> You can fix this by running the following code:

Reviewer 2: Not clear on what they are fixing. The fact that it broke, or how to make a copy of the folder rather than moving it?

Author: I think it's fixing the link between the two folders? 

@mmahacek, could you clarify here? :)

Author: Does this look like an okay disclaimer to cover the workaround?

```suggestion
You can prevent this by running the following code (note that this is a workaround and does not remove the link between the folders):
```

Reviewer 2: ```suggestion
You can fix this by running the following code:
```
Since we've moved the word "fix" I don't think we need to state that it's a workaround.

[source, console]
----
cd /usr/share/opennms
mv etc etc_bad
ln -s /etc/opennms etc
----

Run the following code to move your customized configuration files to `/etc/opennms.bak` and replace them with the contents of `etc-pristine`:

[source, console]
----
mv /etc/opennms /etc/opennms.bak
cp -pR /usr/share/opennms/share/etc-pristine /etc/opennms
----
--

Other OSes::
+
--
Run the following code to move your `$OPENNMS_HOME/etc` files to a safe place, then copy `etc-pristine` to the `$OPENNMS_HOME/etc` directory:

[source, console]
----
mv $OPENNMS_HOME/etc $OPENNMS_HOME/etc.bak
cp -pR $OPENNMS_HOME/share/etc-pristine $OPENNMS_HOME/etc
----

NOTE: On RPM-based installations, the pristine configuration files are located in `$OPENNMS_HOME/share/etc-pristine`.
--
====

> == Initialize git

Author: ```suggestion
== Initialize Git
```

> Run the code below to turn your `etc` directory into a git repository and add the pristine files as your first commit:

Author: ```suggestion
Run the code below to turn your `etc` directory into a Git repository and add the pristine files as your first commit:
```
